jvb/src/main/kotlin/org/jitsi/videobridge/cc/av1/Av1DDAdaptiveSourceProjectionContext.kt

```diff
@@ -0,0 +1,648 @@
+                    frame.isKeyframe = false
+                }
+            }
+            val receivedTime = packetInfo.receivedTime
+            val acceptResult = av1QualityFilter
+                .acceptFrame(frame, incomingEncoding, incomingIndices, targetIndex, receivedTime)
+            frame.isAccepted = acceptResult.accept && frame.index >= lastFrameNumberIndexResumption
+            if (frame.isAccepted) {
+                val projection: Av1DDFrameProjection
+                try {
+                    projection = createProjection(
+                        frame = frame, initialPacket = packet, isResumption = acceptResult.isResumption,
+                        isReset = result.isReset, mark = acceptResult.mark, newDt = acceptResult.newDt,
+                        receivedTime = receivedTime
+                    )
+                } catch (e: Exception) {
+                    logger.warn("Failed to create frame projection", e)
+                    /* Make sure we don't have an accepted frame without a projection in the map. */
+                    frame.isAccepted = false
+                    return false
+                }
+                frame.projection = projection
+                if (projection.earliestProjectedSeqNum isNewerThan lastAv1FrameProjection.latestProjectedSeqNum) {
+                    lastAv1FrameProjection = projection
+                }
+            }
+        }
+
+        val accept = frame.isAccepted && frame.projection?.accept(packet) == true
+
+        if (timeSeriesLogger.isTraceEnabled) {
+            val pt = diagnosticContext.makeTimeSeriesPoint("rtp_av1")
+                .addField("ssrc", packet.ssrc)
+                .addField("timestamp", packet.timestamp)
+                .addField("seq", packet.sequenceNumber)
+                .addField("frameNumber", packet.frameNumber)
+                // TODO add relevant fields from AV1 DD for debugging
+                .addField("targetIndex", Av1DDRtpLayerDesc.indexString(targetIndex))
+                .addField("new_frame", result.isNewFrame)
+                .addField("accept", accept)
+            av1QualityFilter.addDiagnosticContext(pt)
+            timeSeriesLogger.trace(pt)
+        }
+
+        return accept
+    }
+
+    /** Look up an Av1DDFrame for a packet. */
+    private fun lookupAv1Frame(av1Packet: Av1DDPacket): Av1DDFrame? =
+        av1FrameMaps[av1Packet.ssrc]?.findFrame(av1Packet)
+
+    /**
+     * Insert a packet in the appropriate [Av1DDFrameMap].
+     */
+    private fun insertPacketInMap(av1Packet: Av1DDPacket) =
+        av1FrameMaps.getOrPut(av1Packet.ssrc) { Av1DDFrameMap(logger) }.insertPacket(av1Packet)
+
+    private fun haveSubsequentNonAcceptedRequired(
+        frame: Av1DDFrame,
+        incomingEncoding: Int,
+        incomingIndices: Collection<Int>,
+        targetIndex: Int
+    ): Boolean {
+        val map = av1FrameMaps[frame.ssrc] ?: run { return false }
```

Comment:
I don't think you need the `run` here?

Reply:
Without the run we need to be assigning a value to `map`, and the point is to have it be non-null.